src/Features/LanguageServer/Protocol/ExternalAccess/SimplifyTypeNamesHandler.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Simplify;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(SimplifyTypeNamesHandler)), Shared]
+[Method(SimplifyTypeNamesMethodName)]
+internal class SimplifyTypeNamesHandler : ILspServiceDocumentRequestHandler<SimplifyTypeNamesParams, string[]?>
+{
+    private const string SimplifyTypeNamesMethodName = "textDocument/simplifyTypeNames";
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public SimplifyTypeNamesHandler()
+    {
+    }
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(SimplifyTypeNamesParams request) => request.TextDocument;
+
+    public async Task<string[]?> HandleRequestAsync(
+        SimplifyTypeNamesParams request,
+        RequestContext context,
+        CancellationToken cancellationToken)
+    {
+        var document = request.PlacementTextDocument is null
+            ? context.Document
+            : context.Solution!.GetDocument(request.PlacementTextDocument);
+
+        if (document is null)
+            return null;
+
+        var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+        if (semanticModel is null)
+            return null;
+
+        var fullyQualifiedTypeNames = request.FullyQualifiedTypeNames;
+        var result = new string[fullyQualifiedTypeNames.Length];
+        for (var i = 0; i < fullyQualifiedTypeNames.Length; i++)
```

Comment:
I commented in the Razor PR, but wanted to call it out here: https://github.com/dotnet/razor/pull/9070#discussion_r1281273473

I think it would be better to have this endpoint modify the `Document` and insert the code that Razor wants, annotate the nodes, then call the [`Simplifier`](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Simplification/Simplifier.cs,39) and have that do whatever it does. That way we get potential benefits from simplifications other than just type names. It also means in future we could expand this to hook into Roslyns GenerateMethod code to create the method, or have it do things like ensuring visibility modifiers match the users preferences etc.

It could also return that as a `TextEdit[]` too, allowing it to add using declarations if necessary, and allowing us to run it all through our existing formatting engine on the Razor side.

Reply:
Yeah this opens up a lot more potential benefits. Looking into changing the code to do it this way instead.